zebra-consensus/src/checkpoint.rs

```diff
@@ -3,315 +3,1098 @@
+            let parent_hash = qblocks
+                .iter()
+                .skip_while(|qblock| qblock.hash != expected_hash)
+                .map(|qblock| qblock.block.header.previous_block_hash)
+                .next();
+
+            if let Some(parent_hash) = parent_hash {
+                expected_hash = parent_hash;
+                last_height = Some(height);
+            } else {
+                // There was no valid block in the queued block list
+                // Missing blocks are not an error - wait for more blocks
+                //
+                // This check also handles zero-length vecs, which should be
+                // impossible
+                return (Some(expected_height), Some(expected_hash));
+            }
+        }
+
+        // If the loop didn't execute, we're waiting for the next checkpoint block
+        // TODO(teor): add tests for this case
+        if last_height.is_none() {
+            return (Some(next_checkpoint_height), Some(next_checkpoint_hash));
+        }
+
+        // At this point, last_height is the height of the last block, but
+        // expected_hash is the hash of the last block's parent block.
+        //
+
+        // If we're at the genesis block, we want to return None for the parent
+        // block height.
+        let parent_height = last_height.unwrap().0.checked_sub(1).map(BlockHeight);
+        (parent_height, Some(expected_hash))
+    }
+
+    /// Check all the blocks in the current checkpoint range. Send `Ok` for the
+    /// blocks that are in the chain, and `Err` for side-chain blocks.
+    ///
+    /// Does nothing if verification has finished.
+    ///
+    /// Check that the current checkpoint range is complete before calling this
+    /// function.
+    fn submit_current_checkpoint_range(&mut self) {
+        // If checkpoint verification has finished, there should be no queued blocks.
+        if self.current_checkpoint_range.is_none() {
+            return;
+        }
+
+        let next_checkpoint_height = self.get_next_checkpoint_height().unwrap();
+        let current_range = self.current_checkpoint_range.unwrap();
+
+        // Verify all the blocks and discard all the bad blocks in the current range.
+        let mut expected_hash = self.get_next_checkpoint_hash().unwrap();
+        let qrange = self.queued.range_mut(current_range).rev();
+        for (_, qblocks) in qrange {
+            // Find a queued block at each height that is part of the hash chain
+            //
+            // There are two possible outcomes here:
+            //   - there is one block, and it matches the chain (the common case)
+            //   - there are multiple blocks, and at least one block matches the chain
+            //     (if there are duplicate blocks, one succeeds, and the others fail)
+
+            // The caller should check for a continuous chain.
+            debug_assert!(!qblocks.is_empty());
```

Comment:
we should avoid using debug_asserts
```suggestion
            assert!(!qblocks.is_empty());
```

Reply:
Fixed in 42d06094a077bc84e97a8f866cdcf49efb716d3b